java-runtime/java_integration_test.go

```diff
@@ -8,51 +8,158 @@ import (
-			message, ok := response["time"].(string)
-			assert.True(t, ok, "time is not a string: %s", repr.String(response))
-			result, err := time.Parse(time.RFC3339, message)
-			assert.NoError(t, err, "time is not a valid RFC3339 time: %s", message)
-			assert.True(t, result.After(time.Now().Add(-time.Minute)), "time is not recent: %s", message)
-		}),
-		// We call both the go and pass through Java versions
-		// To make sure the response is the same
-		in.Call("gomodule", "emptyVerb", in.Obj{}, func(t testing.TB, response in.Obj) {
-			assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-		}),
-		in.Call("javamodule", "emptyVerb", in.Obj{}, func(t testing.TB, response in.Obj) {
-			assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-		}),
-		in.Call("gomodule", "sinkVerb", "ignored", func(t testing.TB, response in.Obj) {
-			assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-		}),
-		in.Call("javamodule", "sinkVerb", "ignored", func(t testing.TB, response in.Obj) {
-			assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-		}),
-		in.Call("gomodule", "sourceVerb", in.Obj{}, func(t testing.TB, response string) {
-			assert.Equal(t, "Source Verb", response, "expecting empty response, got %s", response)
-		}),
-		in.Call("javamodule", "sourceVerb", in.Obj{}, func(t testing.TB, response string) {
-			assert.Equal(t, "Source Verb", response, "expecting empty response, got %s", response)
-		}),
-		in.Fail(
-			in.Call("gomodule", "errorEmptyVerb", in.Obj{}, func(t testing.TB, response in.Obj) {
-				assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-			}), "verb failed"),
-		in.Fail(
-			in.Call("gomodule", "errorEmptyVerb", in.Obj{}, func(t testing.TB, response in.Obj) {
-				assert.Equal(t, map[string]any{}, response, "expecting empty response, got %s", repr.String(response))
-			}), "verb failed"),
+		in.SubTests(tests...),
 	)
 }
+
+func PairedTest(name string, testFunc func(module string) in.Action) []in.SubTest {
+	return []in.SubTest{
+		{
+			Name:   name + "-go",
+			Action: testFunc("gomodule"),
+		},
+		{
+			Name:   name + "-Java",
+			Action: testFunc("javamodule"),
+		},
+	}
+}
+
+func VerbTest[T any](verb string, value T) func(module string) in.Action {
+	return func(module string) in.Action {
+		return in.Call(module, verb, value, func(t testing.TB, response T) {
+			assert.Equal(t, value, response, "verb call results not equal %s %s", value, response)
+		})
+	}
+}
+
+func PairedVerbTest[T any](verb string, value T) []in.SubTest {
+	return PairedTest(verb, VerbTest[T](verb, value))
+}
+
+func PairedPrefixVerbTest[T any](prefex string, verb string, value T) []in.SubTest {
+	return PairedTest(prefex+"-"+verb, VerbTest[T](verb, value))
+}
+
+type TestObject struct {
```

Comment:
We already have an approach in place for doing multi-language tests, I'd _really_ prefer we don't recreate that in a bespoke manner for the Java runtime.

To make that work with the existing system we'll need some mechanism to tell the integration testing system which languages to run the same test against. I'll send a PR to support this.

Reply:
Won't chain fail at the first failure? As I had a large number of tests it was really useful to just see the overall state in the IDE.

I didn't realize there was already this sort of multi language testing, I should have checked first.

Reply:
Also `TestObject` is just an arbitrary object with lots of different field types to test how they serialize, it is not really part of the testing system.

Reply:
Yeah I understand that `TestObject` is common to both languages, this kind of "conformance testing" to ensure that the behaviour across multiple languages is identical is exactly what the integration framework was designed for. We actually used to have 1:1 integration test coverage across kotlin and go, but sadly we had to delete them when we paused support :(